pkg/controller/apmserverelasticsearchassociation/apmserverelasticsearchassociation_controller.go

```diff
@@ -239,28 +240,46 @@ func (r *ReconcileApmServerElasticsearchAssociation) isCompatible(ctx context.Co
 }
 
 func (r *ReconcileApmServerElasticsearchAssociation) reconcileInternal(ctx context.Context, apmServer *apmv1.ApmServer) (commonv1.AssociationStatus, error) {
+	// garbage collect leftover resources that are not required anymore
+	if err := deleteOrphanedResources(ctx, r, apmServer); err != nil {
+		log.Error(err, "Error while trying to delete orphaned resources. Continuing.", "namespace", apmServer.Namespace, "as_name", apmServer.Name)
+	}
+	apmServerKey := k8s.ExtractNamespacedName(apmServer)
 	// no auto-association nothing to do
 	elasticsearchRef := apmServer.Spec.ElasticsearchRef
 	if !elasticsearchRef.IsDefined() {
-		return commonv1.AssociationUnknown, nil
+		// stop watching any ES cluster previously referenced for this Kibana resource
+		r.watches.ElasticsearchClusters.RemoveHandlerForKey(elasticsearchWatchName(apmServerKey))
```

Comment:
Should we also remove the watch attached to the user for the previously referenced ES secret? (eg. `r.watches.Secrets.RemoveHandlerForKey(elasticsearchWatchName(obj))`)
It's tricky since we don't have that reference anymore.

Reply:
The watch name does not depend on the Es name.
I kept the watch on the Secrets on purpose here. It does not cost that much and it has the benefit to delete any unnecessary secret that would be created "externally" (e.g. by the end user or restored from a backup). A reconciliation would occur and they would be deleted by `deleteOrphanedResources`

All the watchers are cleaned during the `onDelete` call.

Reply:
Discussed out of band: this is an implicit mechanism that would be hard to understand later. We reuse the `onDelete` function to clean up the watches instead.